langchain/src/vectorstores/elasticsearch.ts

```diff
@@ -0,0 +1,212 @@
+import { Client, estypes } from "@elastic/elasticsearch";
+import { Embeddings } from "../embeddings/base.js";
+import { Document } from "../document.js";
+import { VectorStore } from "./base.js";
+
+type ElasticKnnEngine = "hnsw";
+type ElasticSimilarity = "l2_norm" | "dot_product" | "cosine";
+
+interface VectorSearchOptions {
+  readonly engine?: ElasticKnnEngine;
+  readonly similarity?: ElasticSimilarity;
+  readonly m?: number;
+  readonly efConstruction?: number;
+  readonly candidates?: number;
+}
+
+export interface ElasticClientArgs {
+  readonly client: Client;
+  readonly indexName?: string;
+  readonly vectorSearchOptions?: VectorSearchOptions;
+}
+
+type ElasticFilter = object;
+
+export class ElasticVectorSearch extends VectorStore {
+  declare FilterType: ElasticFilter;
+
+  private readonly client: Client;
+
+  private readonly indexName: string;
+
+  private readonly engine: ElasticKnnEngine;
+
+  private readonly similarity: ElasticSimilarity;
+
+  private readonly efConstruction: number;
+
+  private readonly m: number;
+
+  private readonly candidates: number;
+
+  constructor(embeddings: Embeddings, args: ElasticClientArgs) {
+    super(embeddings, args);
+
+    this.engine = args.vectorSearchOptions?.engine ?? "hnsw";
+    this.similarity = args.vectorSearchOptions?.similarity ?? "l2_norm";
+    this.m = args.vectorSearchOptions?.m ?? 16;
+    this.efConstruction = args.vectorSearchOptions?.efConstruction ?? 100;
+    this.candidates = args.vectorSearchOptions?.candidates ?? 200;
+
+    this.client = args.client;
+    this.indexName = args.indexName ?? "documents";
+  }
+
+  async addDocuments(documents: Document[]): Promise<void> {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    return this.addVectors(
+      await this.embeddings.embedDocuments(texts),
+      documents
+    );
+  }
+
+  async addVectors(vectors: number[][], documents: Document[]): Promise<void> {
+    await this.ensureIndexExists(
+      vectors[0].length,
+      this.engine,
+      this.similarity,
+      this.efConstruction,
+      this.m
+    );
+    const operations = vectors.flatMap((embedding, idx) => [
+      {
+        index: {
+          _index: this.indexName,
+        },
+      },
+      {
+        embedding,
+        metadata: documents[idx].metadata,
+        text: documents[idx].pageContent,
+      },
+    ]);
+    await this.client.bulk({ refresh: true, operations });
```

Comment:
this will not be aware to outside if something error happened.
cause they will return something like this

```js
[
 {
    status: 400,
    error: '{\n' +
      '  "type": "mapper_parsing_exception",\n' +
      '  "reason": "failed to parse",\n' +
      '  "caused_by": {\n' +
      '    "type": "illegal_argument_exception",\n' +
      '    "reason": "cannot index vectors with dimension greater than 1024"\n' +
      '  }\n' +
      '}'
  }
]
```

but the result is dropped